staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -402,3 +402,45 @@ func TestUpdate(t *testing.T) {
 	testDoneWG.Wait()
 	close(stop)
 }
+
+func TestPanicPropagated(t *testing.T) {
+	// source simulates an apiserver object endpoint.
+	source := fcache.NewFakeControllerSource()
+
+	// Make a controller that just panic if the AddFunc is called.
+	_, controller := NewInformer(
+		source,
+		&v1.Pod{},
+		time.Millisecond*100,
+		ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				// Create a panic.
+				panic("Just panic.")
+			},
+		},
+	)
+
+	// Run the controller and run it until we close stop.
+	stop := make(chan struct{})
+	defer close(stop)
+
+	propagated := make(chan interface{})
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				propagated <- r
+			}
+		}()
+		controller.Run(stop)
+	}()
+	// Let's add a object to the source. It will trigger a panic.
+	source.Add(&v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "test"}})
+
+	// Check if the panic propagated up.
+	select {
+	case <-propagated:
+		t.Logf("Panic propagated.")
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Errorf("Timeout: the panic stucked!")
```

Comment:
"timeout:  the panic failed to propagate from the controller run method!"

Reply:
Might be worth checking the panic is the one you expect.

Reply:
> Might be worth checking the panic is the one you expect.

Do you mean this code :

```
	var wg wait.Group
	defer func() {
		if r := recover(); r != nil {
			panic(r)
		}
		wg.Wait()
	}()
	wg.StartWithChannel(stopCh, r.Run)

	wait.Until(c.processLoop, time.Second, stopCh)
```

I think it's conciser to call `wg.Wait()` at the end than to check panic in the defer func. And in this case (there is only one defer), it wouldn't change the code ordering actually. Recovering and checking the panic in the defer func would make the stack trace more complex and harder to find out the 'real' panicked code, it may print redundancy information.

Reply:
Thanks for your detailed reply. I got it.